agent/eni/manager.go

```diff
@@ -0,0 +1,242 @@
+// permissions and limitations under the License.
+
+package eni
+
+import (
+	"fmt"
+	"sync"
+	"time"
+	"os"
+	"strings"
+	"context"
+
+	"path/filepath"
+
+	"github.com/fsnotify/fsnotify"
+	"github.com/vishvananda/netlink"
+
+	log "github.com/cihub/seelog"
+)
+
+const (
+	sysfsNetDir			= "/sys/class/net"
+	eth				= "eth"
+	defaultReconciliationInterval	= time.Second * 10
+)
+
+type ENIManager interface {
+	//TODO: Add Helper for payload handler
+	InitENIStateManager() error
+	BeginENIUpdate (ctx context.Context)
+}
+
+
+type ENIStateManager struct {
+	updateLock			sync.RWMutex
+	updateIntervalTicker		*time.Ticker
+	enis				map[string]string // MAC => Device-Name
+	watcher				*fsnotify.Watcher
+}
+
+
+func NewENIManager() ENIManager {
+	return &ENIStateManager{
+		enis: make(map[string]string, 10),
+	}
+}
+
+
+func (eniStateManager *ENIStateManager) InitENIStateManager() error {
+	links, err := netlink.LinkList()
+	if err != nil {
+		log.Errorf("Error retrieving network interfaces: %s", err.Error())
+		return err
+	}
+
+	for _, link := range links {
+		deviceName, MACAddress := link.Attrs().Name, link.Attrs().HardwareAddr.String()
+		if strings.HasPrefix(deviceName, eth) {
+			eniStateManager.addDeviceWithMACAddress(deviceName, MACAddress)
+		}
+	}
+
+	// Setup FSNotify Watcher
+	eniStateManager.watcher, err = fsnotify.NewWatcher()
+	if err != nil {
+		log.Errorf("Error creating watcher: %s", err.Error())
+		return err
+	}
+	// Add Watch Directory
+	err = eniStateManager.watcher.Add(sysfsNetDir)
+	if err != nil {
+		log.Errorf("Error adding watcher: %s", err.Error())
+		return err
+	}
+
+	// FSNotify Update Handler
+	go func () {
+		for {
+			select {
+			case evt := <-eniStateManager.watcher.Events:
+				if evt.Op&fsnotify.Create == fsnotify.Create {
+					eniStateManager.addDevice(evt.Name)
+				}
+				if evt.Op&fsnotify.Remove == fsnotify.Remove {
+					eniStateManager.removeDevice(evt.Name)
+				}
+			case erx := <-eniStateManager.watcher.Errors:
+				log.Debugf("FSNotify Error: %s", erx.Error())
+			}
+		}
+	}()
+
+	return nil
+}
+
+
+func (eniStateManager *ENIStateManager) BeginENIUpdate (ctx context.Context) {
+	eniStateManager.performPeriodicReconciliation(ctx, defaultReconciliationInterval)
+}
+
+
+func (eniStateManager *ENIStateManager) performPeriodicReconciliation (ctx context.Context, updateInterval time.Duration) {
+	eniStateManager.updateIntervalTicker = time.NewTicker(updateInterval)
+	for {
+		select {
+		case <-eniStateManager.updateIntervalTicker.C:
+			go eniStateManager.reconcileENIs()
+		case <-ctx.Done():
+			eniStateManager.updateIntervalTicker.Stop()
+			return
+		}
+	}
+}
+
+
+func (eniStateManager *ENIStateManager) reconcileENIs () {
+	links, err := netlink.LinkList()
+	if err != nil {
+		log.Error("Error obtaining netlink linklist: %s", err.Error())
+	}
+
+	currentState := eniStateManager.buildState(links)
+
+	// Remove non-existent interfaces first
+	for mac, _ := range eniStateManager.enis {
```

Comment:
There may be a race condition here: if another go routine triggered by the event is adding device to the map, where you are removing all of the entries in the map. Maybe you should put lock here instead of the add/remove function.